build/lpc55pins/src/lib.rs

```diff
@@ -0,0 +1,156 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use anyhow::Result;
+use proc_macro2::TokenStream;
+use quote::{format_ident, ToTokens, TokenStreamExt};
+use serde::Deserialize;
+use std::io::Write;
+
+#[derive(Clone, Debug, Deserialize)]
+struct Pin {
+    port: usize,
+    pin: usize,
+}
+
+impl Pin {
+    fn get_port_pin(&self) -> (usize, usize) {
+        if self.pin > 32 {
+            panic!("Invalid pin {}", self.pin);
+        }
+
+        if self.port > 1 {
+            panic!("Invalid port {}", self.port);
+        }
+
+        (self.port, self.pin)
+    }
+}
+
+impl ToTokens for Pin {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let (port, pin) = self.get_port_pin();
+        let final_pin = format_ident!("PIO{}_{}", port, pin);
+        tokens.append_all(quote::quote! {
+            // Yes we want the trailing comma
+            Pin::#final_pin,
+        });
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct PinConfig {
+    pin: Pin,
+    alt: usize,
+    mode: Option<String>,
```

Comment:
Handwavey question – would it make sense to use strongly-typed enums here, instead of `String`?

I ask because we're validate the pin / port numbers in this build script, rather than letting them fail when building the application itself, but doing the opposite for these values (where they pass through codegen just fine, but would then fail during the application compilation)..

Reply:
I hadn't thought about enums because I didn't know if they were possible. That's a good idea.

Reply:
After taking a quick look I'm not sure it's worth it. I found this comment in the net build crate:
```
/// TODO: this type really wants to be an enum, but the toml crate's enum
/// handling is really, really fragile, and currently it would be an enum with a
/// single variant anyway.
```
So it sounds like enum handling may be questionable. Even if we did add enum support, we're currently using these with `format_ident` as the name of a field in an existing enum. Maybe there's a better way to use this with `quote` but if not we'd be having to go back and forth to a string anyway. The current failure mode of the existing code will be at compilation time but it should be pretty easy to determine the issue since it will complain about a mismatch of an enum field. If I'm mistaken about the level of annoyance of doing this I'd appreciate pointers.

Reply:
I haven't tried TOML enums, so probably not worth digging into right now (and failing slightly-later-but-still-at-compile-time is fine by me)